### ISSUE TITLE
unaligned access on mips64 processors (cavium octeon II)

### ISSUE THREAD

**Issue by BrainSlayer** [vel21ripn/nDPI#68]
some of the structures in the ndpi netfilter kernel module (flow tree) are using packed structures. normally this isnt a problem since the compiler handles that situation with some performance decrease of cause. however functions like "set_bit" are implemented in assembly language and will cause a unaligned access which crashes the kernel. i was able to fix that problem by removing the packed attributes from main.c and ndpi_main_netfilter.h i know its mainly just related to the flags in nf_ct_ext_ndpi but for performance reasons it makes no sense at all on this cpu architecture even if it saves some space

here a sample of a kernel log. 

> [ 2042.344203] CPU: 1 PID: 0 Comm: swapper/1 Tainted: P                4.9.192 #1452
[ 2042.357165] task: 800000008f4d4f80 task.stack: 800000008f4d8000
[ 2042.363091] $ 0   : 0000000000000000 0000000010109ce1 ffffffffffffffff 00000000001fff00
[ 2042.371161] $ 4   : 0000000000000300 ffffffff816b0000 ffffffffc0728ccc 0000000000000000
[ 2042.379230] $ 8   : 800000008f57a000 800000008f413aa8 800000008f413aa8 0000000000000000
[ 2042.387299] $12   : 0000000000000028 ffffffff812d7e68 0000000000000000 0000000000000000
[ 2042.395367] $16   : 8000000087121ac8 8000000087121b6c 80000000873e8000 ffffffffc07b0000
[ 2042.403436] $20   : ffffffff8150cae8 8000000087121b78 0000000002080020 0000000000000800
[ 2042.411505] $24   : 0020000000000000 ffffffff8144abe0
[ 2042.419573] $28   : 800000008f4d8000 800000008f413c20 800000008f57a0a0 ffffffffc077bec4
[ 2042.427644] Hi    : 0000000000000000
[ 2042.431222] Lo    : 0000000000000000
[ 2042.434846] epc   : ffffffffc077bec8 nf_ndpi_free_flow+0x1e8/0x388 [xt_ndpi]
[ 2042.441924] ra    : ffffffffc077bec4 nf_ndpi_free_flow+0x1e4/0x388 [xt_ndpi]
[ 2042.448980] Status: 10109ce3 KX SX UX KERNEL EXL IE
[ 2042.454024] Cause : 00800010 (ExcCode 04)
[ 2042.458038] BadVA : 8000000087121b6c
[ 2042.461618] PrId  : 000d9301 (Cavium Octeon II)
[ 2042.466152] Modules linked in: xt_physdev xt_IMQ imq xt_ndpi nf_nat_pptp nf_conntrack_pptp nf_nat_proto_gre nf_conntrack_proto_gre eoip cmac md4 cifs fscache l2tp_ppp l2tp_netlink l2tp_c
ore ip6_udp_tunnel udp_tunnel ipv6 n_hdlc nf_nat_h323 nf_conntrack_h323 gre etherip tun jffs2 lzma_decompress(P) lzma_compress(P) ath9k ath mac80211 compat ath_pci ath_hal(P) tmp421 hwmon b
trfs xor raid6_pq zstd_decompress nls_utf8 zstd_compress nls_cp437 xxhash lzo_decompress lzo_compress [last unloaded: fast_classifier]
[ 2042.511409] Process swapper/1 (pid: 0, threadinfo=800000008f4d8000, task=800000008f4d4f80, tls=0000000000000000)
[ 2042.521595] Stack : 800000008f57a0a0 ffffffff826e10a8 800000008558f420 800000008e106780
[ 2042.529664]         ffffffff826e10b0 800000008f57a078 0000000000000000 ffffffff8144cc3c
[ 2042.537733]         0000000000000003 800000008e106780 ffffffff81658480 800000008f57a000
[ 2042.545802]         fffffffffffffffe ffffffff81442d48 ffffffff815d82e8 8000000087363000
[ 2042.553870]         ffffffff815d82e8 ffffffff8143db90 8000000087363000 ffffffff813dada4
[ 2042.561939]         800000008d75cb54 8000000087363000 800000008d75cb54 ffffffff813dc9f4
[ 2042.570007]         8000000087363000 ffffffff813dcaac 8000000087363000 ffffffff814651f0
[ 2042.578075]         800000008f57a000 8000000087363000 800000008f57a000 ffffffff81658480
[ 2042.586144]         800000008f57a000 ffffffff814659dc 0000000080000000 02ffffff816a5f00
[ 2042.594213]         800000008f57a000 0000000000000000 0000000000000000 ffffffff81658480
[ 2042.602281]         ...
[ 2042.604751] Call Trace:
[ 2042.607223] [<ffffffffc077bec8>] nf_ndpi_free_flow+0x1e8/0x388 [xt_ndpi]
[ 2042.613959] [<ffffffff8144cc3c>] __nf_ct_ext_destroy+0x5c/0x88
[ 2042.619805] [<ffffffff81442d48>] nf_conntrack_free+0x28/0x80
[ 2042.625474] [<ffffffff8143db90>] nf_conntrack_destroy+0x28/0x38
[ 2042.631407] [<ffffffff813dada4>] skb_release_head_state+0x134/0x190
[ 2042.637685] [<ffffffff813dc9f4>] skb_release_all+0x14/0x40
[ 2042.643179] [<ffffffff813dcaac>] kfree_skb+0x64/0x90
[ 2042.648157] [<ffffffff814651f0>] ip_rcv_finish+0x318/0x370
[ 2042.653652] [<ffffffff814659dc>] ip_rcv+0x2e4/0x320
[ 2042.658544] [<ffffffff813f2538>] __netif_receive_skb_core+0xad8/0xc18
[ 2042.664996] [<ffffffff813f3744>] process_backlog+0x94/0x160
[ 2042.670580] [<ffffffff813f2dd8>] net_rx_action+0x128/0x2a8
[ 2042.676075] [<ffffffff81147288>] __do_softirq+0x150/0x300
[ 2042.681483] [<ffffffff81147730>] irq_exit+0xb0/0xd0
[ 2042.686372] [<ffffffff8111546c>] plat_irq_dispatch+0x4c/0x118
[ 2042.692128] [<ffffffff8110a1cc>] handle_int+0x14c/0x158
[ 2042.697361] [<ffffffff8110a040>] __r4k_wait+0x20/0x40
[ 2042.702423] [<ffffffff8117a908>] cpu_startup_entry+0xc8/0x110
[ 2042.708183] Code: 0280f809  661100a4  2402ffff <d2230000> 7c430847  f2230000  1060fffc  00000000  3c02c07b
[ 2042.718049]
[ 2042.719609] ---[ end trace a31d1ffd9ddad117 ]---

>

**Comment by vel21ripn** [comment]
I know what "unaligned access" has been since PDP-11. 
How quickly do oops occur? When processing the first network packet or after some random amount of time?
Stupidly removing the "packed" attribute is not the best solution.
I need to understand exactly where the error occurs. Can you send the result 
```
objdump -l -d main.o
```
starting from the line "<nf_ndpi_free_flow>:" and then to the end of the function (about 300 lines)?

**Comment by BrainSlayer** [comment]
Am 12.09.2019 um 14:17 schrieb Vitaly Lavrov:
>
> I know what "unaligned access" has been since PDP-11.
> How quickly do oops occur? When processing the first network packet or 
> after some random amount of time?
>
immediatly. but it seem to be related to the first packet. so if i 
initialize it and add one rule to the mangle table it crashes 
immediatly. i also tried it on a passive interface. nothing happend
but once i start a single packet it crashes.
>
> Stupidly removing the "packed" attribute is not the best solution.
> I need to understand exactly where the error occurs. Can you send the 
> result
>
my disassembly say its the "set_bit" function. so nf_ct_ext_ndpi 
structure -> flags is the only possible cause. but i wanted to avoid 
other crashes on other locations

since packed structures are a bad thing on mips at all

> |objdump -l -d main.o |
>
> starting from the line "<nf_ndpi_free_flow>:" and then to the end of 
> the function (about 300 lines)?
>
i did this already. as i said. "set_bit" is the cause. i'm experience 
with such debugging :-)


i attach you the disassembly with source comments. look at address 
0x61ec8 from the function. its the crash pointer

**Comment by BrainSlayer** [comment]
[main.zip](https://github.com/vel21ripn/nDPI/files/3605658/main.zip)

**Comment by BrainSlayer** [comment]
consider that this is a modified variant from your code. i included all c files as single compile object which reduces the total size by 250 kb since unused functions can be removed if declared as static. since i work on embedded devices this way is usefull. but the code itself is untouched in that assembly. i made the changes i proposed later and it fixed the unaligned crash

**Comment by vel21ripn** [comment]
> immediatly. but it seem to be related to the first packet. so if i initialize it and add one rule to the mangle table it crashes immediatly. i also tried it on a passive interface. nothing happend but once i start a single packet it crashes.

If the crash causes the first packet, then why do we see nf_ndpi_free_flow() ? Do you use "icmp" for testing? For udp/tcp, nf_ndpi_free_flow() will be called much later.

> my disassembly say its the "set_bit" function. so nf_ct_ext_ndpi structure -> flags is the only possible cause. but i wanted to avoid other crashes on other locations 

 and 

> i attach you the disassembly with source comments. look at address
0x61ec8 from the function. its the crash pointer

main.asm : 0x61ec8  is raw_spin_unlock_bh()

Something is wrong. Of interest is only the assembler code of the variant that causes the crash.

> i included all c files as single compile object which reduces the total size by 250 kb since unused functions can be removed if declared as static.

What tools are used to combine all the ".c" files?

Returning to the problem of aligning fields in structures.
If the problem is only in the "nf_ct_ext_ndpi.flags" field, then it can be solved in another way. The field has an excess length, it can be reduced to uint16_t and moved between the "lock" field and the "l4_proto" field in order to have the correct alignment.
I am currently testing changes made to correctly read data from proto, ip_proto, host_proto and flows files. After that, I can tackle the problem of aligned fields in structures.

**Comment by BrainSlayer** [comment]
Am 12.09.2019 um 20:05 schrieb Vitaly Lavrov:
>
>     immediatly. but it seem to be related to the first packet. so if i
>     initialize it and add one rule to the mangle table it crashes
>     immediatly. i also tried it on a passive interface. nothing
>     happend but once i start a single packet it crashes.
>
> If the crash causes the first packet, then why do we see 
> nf_ndpi_free_flow() ? Do you use "icmp" for testing? For udp/tcp, 
> nf_ndpi_free_flow() will be called much later.
>
standard traffic. nothing synthetic. http, ssh, etc.
>
>     my disassembly say its the "set_bit" function. so nf_ct_ext_ndpi
>     structure -> flags is the only possible cause. but i wanted to
>     avoid other crashes on other locations
>
> and
>
>     i attach you the disassembly with source comments. look at address
>     0x61ec8 from the function. its the crash pointer
>
> main.asm : 0x61ec8 is raw_spin_unlock_bh()
>
look at the assembly in deeper detail. its a indirect use of set_bit. 
you need to understand mips assembly of course. (hint the while(!temp)  
line is only present in set_bit function. same for volatile asm {. )
the line comments arent in correct order always since the compiler 
reorders code while optimization.
>
> Something is wrong. Of interest is only the assembler code of the 
> variant that causes the crash.
>
>     i included all c files as single compile object which reduces the
>     total size by 250 kb since unused functions can be removed if
>     declared as static.
>
> What tools are used to combine all the ".c" files?
>
my hands and fingers. simply include all c files on the main service 
file. edit headers with function declarations to set all functions 
static and correct all errors which may happen.
but i used a macro for static here. so i can manually disable static 
functions for debugging.

for example (just a shorted version)

#include "ndpi_strcol.c"
#include "ndpi_proc_parsers.c"
#include "ndpi_proc_generic.c"
#include "ndpi_proc_info.c"
#include "ndpi_proc_flow.c"
#include "ndpi_proc_hostdef.c"
#include "ndpi_proc_ipdef.c"
#include "../libre/regexp.c"
#include "../../src/lib/ndpi_main.c"

for full source. take a look into my svn server (svn.dd-wrt.com at 
src/router/ndpi-netfilter)


> Returning to the problem of aligning fields in structures.
> If the problem is only in the "nf_ct_ext_ndpi.flags" field, then it 
> can be solved in another way. The field has an excess length, it can 
> be reduced to uint16_t and moved between the "lock" field and the 
> "l4_proto" field in order to have the correct alignment.
> I am currently testing changes made to correctly read data from proto, 
> ip_proto, host_proto and flows files. After that, I can tackle the 
> problem of aligned fields in structures.
>
i changed this already in my source in that way, but you should also 
consider that packed fields need the compiler to handle each field 
access very special. it just makes performance bad on heavily used 
structures.
a x64/x86 handles this easier since the instructions are able to handle 
unaligned access by themself, but performance decrease is present here too.
if not really neccessary like on ip header structures where it is 
required that the alignment follows the dataformat. never use packed 
structures. its bad and makes code all slower.
guaranteed.
its common to convert such dataformat structures to internal aligned 
structures before processing them for optimizing performance. but this 
is all not required here on these internal structures.

**Comment by vel21ripn** [comment]
It is unfortunate that you are using svn. This complicates the process of making useful changes o ndpi-netfilter.
Right "packing" of the fields allows you to get the minimum length of the structure without loss of performance. I found the "pahole" utility that simplifies the process of packing fields in a data structure.
Now I am analyzing the need for atomic operations such as set_bit()/clear_bit() for a "flags" field.
Perhaps they can be replaced with regular bit operations, which are much faster.
The use of atomic bit operations was introduced at the stage of combating errors that occurred during competitive data access. After fixing these errors, atomic bit operations were not replaced by ordinary bit operations.

**Comment by BrainSlayer** [comment]
no packing decreases performance since it will create unaligned accesses which decreases the performance on all sorts of cpu types.  about atomic operations. these are required on time critical paths like interrupt contexts or special even handling, but nowhere else. but replacing it with a regular implementation will still have a unaligned access. the compiler works just around it and this creates performances losses. and this is the point where is say.  get rid of it. its not neccessary. you save a few bytes but loose 30% of performance. also on a x86/x64 cpu will loose performance. so please read carefull how cpu's are designed and why aligned accesses are important on any cpu type on this planet

**Comment by BrainSlayer** [comment]
svn isnt complicated. for me its historic use. i need to track revision numbers. 32 byte hashes are just stupid and tells me nothing about the version. git is too complicated to be usefull for me. if i want to compare projects i just use the standard diff command. it helps me to review changes with any standard diff browser. consider that i only work on commandline. no fancy gui utils etc. and on commandline git is just shit

**Comment by vel21ripn** [comment]
Look at the commits cff4144 and 219bb9a.
I saved the "packed" attribute only where it really is needed.
I changed the order of the fields in some structures to get the correct alignment of the data in the structures and refused atomic operations with the flags field.
The main project has accumulated a lot of changes that need to be applied in ndpi-netfilter, so in the near future many changes will appear.
